src/QsCompiler/CompilerOptimizations/TransformationState.fs

```diff
@@ -0,0 +1,104 @@
+    currentCallable: QsCallable option
+}
+
+/// Creates a new TransformationState with the given compiledCallables
+let internal newState compiledCallables =
+    {compiledCallables = compiledCallables; scopeStack = []; currentCallable = None}
+
+
+/// Gets the QsCallable with the given qualified name
+let internal getCallable state qualName =
+    state.compiledCallables.[qualName]
+
+
+/// Returns a TransformationState inside of a new scope
+let internal enterScope state =
+    {state with scopeStack = Map.empty :: state.scopeStack}
+
+/// Returns a TransformationState outside of the current scope
+let internal exitScope state =
+    match state.scopeStack with
+    | _ :: tail -> {state with scopeStack = tail}
+    | [] -> failwithf "No scope to exit"
+
+/// Gets the current value of the given variable
+let internal getVar state name =
+    state.scopeStack |> List.tryPick (Map.tryFind name)
+    
+
+/// Returns whether a given expression is a literal (and thus a constant)
+let rec internal isLiteral (state: TransformationState) (expr: Expr): bool =
+    match expr with
+    | UnitValue | IntLiteral _ | BigIntLiteral _ | DoubleLiteral _ | BoolLiteral _ | ResultLiteral _ | PauliLiteral _ -> true
+    | ValueTuple a | StringLiteral (_, a) | ValueArray a -> Seq.forall (fun x -> isLiteral state x.Expression) a
+    | RangeLiteral (a, b) -> isLiteral state a.Expression && isLiteral state b.Expression
+    | NewArray (_, a) -> isLiteral state a.Expression
+    | Identifier (GlobalCallable _, _) | MissingExpr -> true
+    | CallLikeExpression ({Expression = Identifier (GlobalCallable qualName, _)}, b)
+        when (getCallable state qualName).Kind = TypeConstructor -> isLiteral state b.Expression
+    | CallLikeExpression (a, b) ->
+        isLiteral state a.Expression && isLiteral state b.Expression &&
+            TypedExpression.IsPartialApplication (CallLikeExpression (a, b))
+    | _ -> false
+
+
+/// Returns a TransformationState with the given variable defined as the given value
+let internal defineVar state (name, value) =
+    if not (isLiteral state value) then state else
+    // TODO: assert variable is undefined
+    match state.scopeStack with
+    | head :: tail ->
+        let newHead = head.Add (name, value)
+        {state with scopeStack = newHead :: tail}
+    | [] -> failwithf "No scope to define variables in"
+
+/// Returns a TransformationState with the given variable set to the given value
+let internal setVar state (name, value) =
+    if not (isLiteral state value) then state else
+    // TODO: assert variable is defined, is same type
+    match state.scopeStack |> List.tryFindIndex (Map.containsKey name) with
+    | Some index ->
+        let newScopeStack = state.scopeStack |> List.mapi (fun i -> if i = index then Map.add name value else id)
+        {state with scopeStack = newScopeStack}
+    | None -> failwithf "Variable %s is undefined" name
+
+/// Applies on of the operations above on a SymbolTuple, ValueTuple pair
```

Comment:
Can you be more descriptive? I think you mean one of the functions in this module rather than a Q# operation?

```suggestion
/// Applies the given function op on a SymbolTuple, ValueTuple pair
```